features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,931 @@
+/*
+ * Copyright (c) 2018 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// ----------------------------------------------------------- Includes -----------------------------------------------------------
+
+#include "nvstore.h"
+
+#if NVSTORE_ENABLED
+
+#include "FlashIAP.h"
+#include "mbed_critical.h"
+#include "mbed_assert.h"
+#ifdef MBED_CONF_RTOS_PRESENT
+#include "Thread.h"
+#endif
+#include <algorithm>
+#include <string.h>
+#include <stdio.h>
+
+// --------------------------------------------------------- Definitions ----------------------------------------------------------
+
+#define DELETE_ITEM_FLAG  0x8000
+#define SET_ONCE_FLAG     0x4000
+#define HEADER_FLAG_MASK  0xF000
+
+#define MASTER_RECORD_KEY 0xFFE
+#define NO_KEY            0xFFF
+#define LAST_RESERVED_KEY MASTER_RECORD_KEY
+
+typedef struct
+{
+    uint16_t key_and_flags;
+    uint16_t size;
+    uint32_t mac;
+} record_header_t;
+
+
+#define OFFS_BY_KEY_AREA_MASK     0x80000000UL
+#define OFFS_BY_KEY_SET_ONCE_MASK 0x40000000UL
+#define OFFS_BY_KEY_FLAG_MASK     0xC0000000UL
+#define OFFS_BY_KEY_AREA_BIT_POS      31
+#define OFFS_BY_KEY_SET_ONCE_BIT_POS  30
+
+#define MASTER_RECORD_BLANK_FIELD_SIZE FLASH_MINIMAL_PROG_UNIT
+
+typedef struct {
+    uint16_t version;
+    uint16_t reserved1;
+    uint32_t reserved2;
+} master_record_data_t;
+
+#define MASTER_RECORD_SIZE sizeof(master_record_data_t)
+
+#define MEDITATE_TIME_MS 1
+
+#define MIN_AREA_SIZE 4096
+
+#define NUM_WRITE_RETRIES 16
```

Comment:
To avoid clashes/redefinition, `NVSTORE_` prefix would be good to have for these macros in this file (although they are local)

Or as this is C++, use internal const variables, `static const uint8_t num_write_retries` for instance

Reply:
Done that (with all other comments). Again squashed due to changes being hard to follow (used the static const notion).